Adds CUDA FP16 arithmetic operators

testhound opened the pull request:
This pull request add operator support for the following fp 16 operations: +, -, *, neg, abs. This PR partly addresses the meta issue "add fp16 support": https://github.com/numba/numba/issues/4402.

This PR builds on: https://github.com/numba/numba/pull/7460 which added intrinsic support for the above operators.

testhound commented:
gpuci run tests

isVoid commented on numba/cuda/cudadecl.py:
```diff
@@ -346,6 +347,19 @@ class Cuda_fp16_unary(ConcreteTemplate):
+        key = l_key
+
+        def generic(self, args, kws):
+            assert not kws
```
Got a small question: `assert` can be manually skipped in production code with `python -O`. Is this something we want to explicitly check for? If so I would suggest check explicitly with `if ...: raise`.

gmarkall replied:
I think this comes from a pattern that already exists a couple of times in `cudadecl`, and it's not something that really adds much value - I've never seen it get hit, and I think it probably just exists because it was perhaps once useful a long time ago (maybe in NumbaPro). I think instead of asserting, we could maybe just `return None` (as in, fail to type) if `kws` is something that's not false-ish.

gmarkall replied:
I think this comes from a pattern that already exists a couple of times in `cudadecl`, and it's not something that really adds much value - I've never seen it get hit, and I think it probably just exists because it was perhaps once useful a long time ago (maybe in NumbaPro). I think instead of asserting, we could maybe just `return None` (as in, fail to type) if `kws` is something that's not false-ish.

isVoid reviewed (commented):
Got a small Q below.

gmarkall commented:
gpuci run tests

testhound commented:
> With this branch, the following:
> 
> ```python
> from numba import cuda
> from numba.types import float16, float32
> 
> def f(x, y):
>     return x + y
> 
> ptx_16, resty_16 = cuda.compile_ptx(f, (float16, float16), device=True)
> ptx_32, resty_32 = cuda.compile_ptx(f, (float32, float32), device=True)
> 
> print(ptx_16)
> print(resty_16)
> 
> print(ptx_32)
> print(resty_32)
> ```
> 
> When run with
> 
> ```
> numba --annotate-html scalar.html scalar.py
> ```
> 
> Seems to give the following typing for the addition of `float16`s (in the HTML output):
> 
> ```
> x = arg(0, name=x)  :: float16
> y = arg(1, name=y)  :: float16
> $6binary_add.2 = x + y  :: float32
> ```
> 
> and corresponding PTX:
> 
> ```
> 	// begin inline asm
> 	cvt.f32.f16 %f1, %rs1;
> 	// end inline asm
> 	// begin inline asm
> 	cvt.f32.f16 %f2, %rs2;
> 	// end inline asm
> 	add.f32 	%f3, %f1, %f2;
> 	st.f32 	[%rd1], %f3;
> ```
> 
> Is it expected with this PR that `float16 + float16` should be typed as returning a `float16`?
> 
> (Does this reproduce for you? Perhaps I am missing something / running something incorrectly?)

@gmarkall  this reproduces for me and you definitely found a issue. Thanks.

isVoid commented on numba/cuda/cudaimpl.py:
```diff
@@ -489,8 +489,14 @@ def ptx_fp16_binary(context, builder, sig, args):
 
 lower_fp16_binary(stubs.fp16.hadd, 'add')
+lower_fp16_binary(operator.add, 'add')
+lower_fp16_binary(operator.iadd, 'add')
```
`iadd` is inplace add "+=". We need to add tests to for all inplace operators. Also, [ptx](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#half-precision-floating-point-instructions) doesn't support `iadd` opcode. We might need to change the result register as the lhs in `lower_fp16_binary` for this feature.

sklam replied:
In python, "inplace" operation `a += b` can mean `a = a + b` for immutable types. All number types are immutable so it is okay to just reuse the implementation for the non-inplace version.

sklam replied:
clarifying: `a = a + b`, the `+` can create a new object which is then bound to the variable `a` while the old object in `a` is not modified.

example:

```python
>>> a = "a"
>>> b = "b"
>>> c = a
>>> a += b
>>> c
'a'
>>> a
'ab'
```

isVoid replied:
TIL, thanks for clarifying!

gmarkall replied:
@isVoid just to check, do you think this is now resolved? (https://github.com/numba/numba/pull/7885/commits/7b36527c847c28fe1abba228753b954f56db539b added some inplace binary tests - what do you think of them)?

gmarkall replied:
I've looked over the tests and have made comments there, so I'll resolve this conversation.

sklam commented on numba/cuda/cudadecl.py:
```diff
@@ -355,11 +369,33 @@ class Cuda_fp16_binary(ConcreteTemplate):
 
 
+def _genfp16_binary_operator(l_key):
+    @register
```
```suggestion
    @register_global
```

testhound replied:
Change made although the key needed to be added as an argument.

sklam commented on numba/cuda/cudadecl.py:
```diff
@@ -346,6 +347,19 @@ class Cuda_fp16_unary(ConcreteTemplate):
 
 
+def _genfp16_unary_operator(l_key):
+    @register
```
```suggestion
    @register_global
```

testhound replied:
Change made although the key needed to be added as an argument.

sklam reviewed (changes requested):
It turns out the template should be using `@register_global` for the `operator.*`. 

You can check it without CUDA hardware with:

```python
import operator
from numba.core import types, typing
from numba.cuda.descriptor import cuda_target

fp16 = types.float16

typingctx = cuda_target.typing_context
typingctx.refresh()
fnty = typingctx.resolve_value_type(operator.add)
out = typingctx.resolve_function_type(fnty, (fp16, fp16), {})
print(out)
if out != typing.signature(fp16, fp16, fp16):
    raise AssertionError(out)
```
The above code is what the compiler does for function resolution. I'd recommend adding tests like this to `numba/cuda/tests/nocuda`.

testhound commented:
gpuci run tests

testhound commented:
gpuci run tests

testhound commented:
gpuci run tests

testhound commented:
gpuci run tests

gmarkall reviewed (changes requested):
I've had a quick pass over this. Apart from the merge from `main` that's needed, there's also a suggestion from @sklam for testing the typing logic by adding tests in the `numba.cuda.tests.nocuda` module that is not yet addressed.

I think that #8010 is probably a bit closer to going in and there will be some commonality in the typing, so we should probably have a look at how this looks on top of that once it's gone in.

testhound commented:
gpuci run tests

testhound commented:
@gmarkall I updated this branch by merging with main and making some changes to test cases. I have queried @sklam with the function resolution test cases are really necessary.

testhound commented:
gpuci run tests

testhound commented:
@gmarkall and @sklam as requested, I added function resolution test cases. @gmarkall I also updated the comments on _genfp16_binary_operator which uses AbstractTemplate to your comment suggestion from the comparsion operator PR.

sklam reviewed (dismissed):
new tests in b2bef05e0490d161ecec07b4403964c19875ee5d looks good

sklam commented:
there's some more conflict due to recent merges

testhound commented:
gpuci run tests

testhound commented:
@sklam  I resolved the conflicts with the main branch.

gmarkall commented on numba/cuda/tests/nocuda/test_function_resolution.py:
```diff
@@ -0,0 +1,35 @@
+
+
+@skip_on_cudasim("Skip on simulator due to use of cuda_target")
+class TestFunctionResolutionNoCuda(unittest.TestCase):
```
This test will error if NVVM is not installed, so it needs skipping in that case.

Also, including `NoCuda` in the class name for a test in the `nocuda` module seems redundant.

```suggestion
@unittest.skipIf(not nvvm.is_available(), "No libNVVM")
class TestFunctionResolution(unittest.TestCase):
```

gmarkall commented on numba/cuda/tests/nocuda/test_function_resolution.py:
```diff
@@ -0,0 +1,35 @@
+from numba.cuda.testing import unittest, skip_on_cudasim
+import operator
+from numba.core import types, typing
```
For suggestion below:

```suggestion
from numba.core import types, typing
from numba.cuda.cudadrv import nvvm
```

testhound commented:
@gmarkall test case updates with your comments.

gmarkall commented:
gpuci run tests

gmarkall commented on numba/cuda/tests/cudapy/test_operator.py:
```diff
@@ -105,6 +137,118 @@ def test_floordiv(self):
+
+                kernel[1, 1](got, arg1[0], arg2[0])
+                expected = op(arg1, arg2)
+                np.testing.assert_allclose(got[0], expected)
```
This compares the scalar first element of `got` with the array `expected`. It would be simpler and clearer to just compare the two arrays.
```suggestion
                np.testing.assert_allclose(got, expected)
```

gmarkall commented on numba/cuda/tests/cudapy/test_operator.py:
```diff
@@ -105,6 +137,118 @@ def test_floordiv(self):
+    @skip_on_cudasim('Compilation unsupported in the simulator')
+    def test_fp16_binary_ptx(self):
+        functions = (simple_fp16add, simple_fp16sub, simple_fp16mul)
+        ops = (operator.add, operator.sub, operator.mul)
+        opstring = ('add.f16', 'sub.f16', 'mul.f16')
+        args = (f2[:], f2, f2)
+
+        for fn, op, s in zip(functions, ops, opstring):
+            with self.subTest(op=op):
+                ptx, _ = compile_ptx(fn, args, cc=(5, 3))
+                self.assertIn(s, ptx)
```
`ops` is not needed here. I think this could be simplified and clarified to:

```suggestion
        instrs = ('add.f16', 'sub.f16', 'mul.f16')
        args = (f2[:], f2, f2) 

        for fn, instr in zip(functions, instrs):
            with self.subTest(instr=instr):
                ptx, _ = compile_ptx(fn, args, cc=(5, 3)) 
                self.assertIn(instr, ptx)
```

gmarkall commented on numba/cuda/tests/cudapy/test_operator.py:
```diff
@@ -105,6 +137,118 @@ def test_floordiv(self):
+
+                arg1 = np.random.random(1).astype(np.float16)
+                arg2 = (np.random.random(1) * 100).astype(ty)
+                arg2_ty = np.result_type(np.float16, ty)
+
+                expected = np.zeros(1, dtype=arg2_ty)
+                got = np.zeros(1, dtype=arg2_ty)
+
+                kernel[1, 1](got, arg1[0], arg2[0])
+                expected = op(arg1, arg2)
+                np.testing.assert_allclose(got[0], expected)
```
Isn't `arg2_ty` holding the type of the result, not the type of `arg2`? We should also compare arrays rather than a scalar and an array, and we don't need to assign `expected` twice:
```suggestion
                res_ty = np.result_type(np.float16, ty) 

                got = np.zeros(1, dtype=res_ty)
                kernel[1, 1](got, arg1[0], arg2[0])
                expected = op(arg1, arg2)

                np.testing.assert_allclose(got, expected)
```

gmarkall commented on numba/cuda/tests/cudapy/test_operator.py:
```diff
@@ -105,6 +137,118 @@ def test_floordiv(self):
+    @skip_on_cudasim('Compilation unsupported in the simulator')
+    def test_fp16_inplace_binary_ptx(self):
+        functions = (simple_fp16_iadd, simple_fp16_isub, simple_fp16_imul)
+        ops = (operator.iadd, operator.isub, operator.imul)
+
+        opstring = ('add.f16', 'sub.f16', 'mul.f16')
+        args = (f2[:], f2)
+
+        for fn, op, s in zip(functions, ops, opstring):
+            with self.subTest(op=op):
+                ptx, _ = compile_ptx(fn, args, cc=(5, 3))
+                self.assertIn(s, ptx)
```
Similar to above, `ops` is not needed here:
```suggestion
        instrs = ('add.f16', 'sub.f16', 'mul.f16')
        args = (f2[:], f2, f2) 

        for fn, instr in zip(functions, instrs):
            with self.subTest(instr=instr):
                ptx, _ = compile_ptx(fn, args, cc=(5, 3)) 
                self.assertIn(instr, ptx)
```

gmarkall commented on numba/cuda/tests/cudapy/test_operator.py:
```diff
@@ -105,6 +137,118 @@ def test_floordiv(self):
+            with self.subTest(op=op):
+                kernel = cuda.jit("void(f2[:], f2)")(fn)
+
+                expected = np.zeros(1, dtype=np.float16)
+                got_in_out = np.random.random(1).astype(np.float16)
+                arg1 = np.random.random(1).astype(np.float16)
+
+                kernel[1, 1](got_in_out, arg1[0])
+                expected = op(got_in_out, arg1)
+                np.testing.assert_allclose(got_in_out[0], expected)
```
As above, we don't need to assign `expected` twice, and should compare arrays.

Additionally, this is modifying `got_in_out` twice, once when we run the kernel and once calling `op`, because these are inplace operations. Instead we should have separate arrays for them:

```suggestion
                got = np.random.random(1).astype(np.float16)
                expected = got.copy()
                arg = np.random.random(1).astype(np.float16)[0]

                kernel[1, 1](got, arg)
                op(expected, arg)
                np.testing.assert_allclose(got, expected)
```

gmarkall commented on numba/cuda/tests/cudapy/test_operator.py:
```diff
@@ -105,6 +137,118 @@ def test_floordiv(self):
+            with self.subTest(op=op):
+                kernel = cuda.jit("void(f2[:], f2)")(fn)
+
+                expected = np.zeros(1, dtype=np.float16)
```
Superfluous assignment:
```suggestion
```

gmarkall commented on numba/cuda/tests/cudapy/test_operator.py:
```diff
@@ -105,6 +137,118 @@ def test_floordiv(self):
+
+                kernel[1, 1](got, arg1[0])
+                expected = op(arg1)
+                np.testing.assert_allclose(got[0], expected)
```
We should compare arrays instead of a scalar and an array:
```suggestion
                np.testing.assert_allclose(got, expected)
```

gmarkall commented on numba/cuda/tests/cudapy/test_operator.py:
```diff
@@ -105,6 +137,118 @@ def test_floordiv(self):
+    def test_fp16_abs_ptx(self):
+        args = (f2[:], f2)
+        ptx, _ = compile_ptx(simple_fp16abs, args, cc=(5, 3))
+        if cuda.runtime.get_version() < (10, 2):
+            self.assertRegex(ptx, r'and\.b16.*0x7FFF;')
+        else:
+            self.assertIn('abs.f16', ptx)
```
We only support CUDA 10.2 onwards now, so the check for prior versions is not needed:
```suggestion
        self.assertIn('abs.f16', ptx)
```

gmarkall commented on numba/cuda/cudadecl.py:
```diff
@@ -355,6 +368,50 @@ class Cuda_fp16_binary(ConcreteTemplate):
+#
+# To work around this, we instead use an AbstractTemplate that implements
+# exactly the casting logic that we desire. The AbstractTemplate gets
+# considered in preference to ConcreteTemplates during typing.
+#
+# This is tracked as Issue #7863 (https://github.com/numba/numba/issues/7863) -
+# once this is resolved it should be possible to replace this AbstractTemplate
+# with a ConcreteTemplate to simplify the logic.
+
+
+def _genfp16_binary_operator(l_key):
+    @register_global(l_key)
+    class Cuda_fp16_binary2(AbstractTemplate):
+        key = l_key
+
+        def generic(self, args, kws):
+            assert not kws
+
+            if len(args) == 2 and \
+                    (args[0] == types.float16 or args[1] == types.float16):
+                if (args[0] == types.float16):
+                    convertible = self.context.can_convert(args[1], args[0])
+                else:
+                    convertible = self.context.can_convert(args[0], args[1])
+
+                # We allow three cases here:
+                #
+                # 1. Comparing fp16 to fp16 - Conversion.exact
+                # 2. Comparing fp16 to types fp16 can be promoted to
+                # - Conversion.promote
+                # 3. Comparing fp16 to int8 (safe conversion) -
+                # Conversion.safe
+                if (convertible == Conversion.exact) or \
+                   (convertible == Conversion.promote) or \
+                   (convertible == Conversion.safe):
+                    return signature(types.float16, types.float16,
+                                     types.float16)
+
+    return Cuda_fp16_binary2
+
```
This appears to be an exact copy/paste of the code and associated comments for `_genfp16_comparison_operator` except for the name of the function and inner class (lines 439-477). Is there something more subtle going on here that I'm not spotting?

gmarkall replied:
After some more inspection of the code, I spotted that they differ in the return type. The point I was trying to make about copy/pasting is that it's probably unnecessary, and it's difficult to maintain large blocks of almost-identical copy/pasted code and comments (for example, it's hard to spot small differences between the variants). The following seems to apply OK on top of your branch and saves duplicating everything about the typing for binary operators and comparisons:

```diff
diff --git a/numba/cuda/cudadecl.py b/numba/cuda/cudadecl.py
index 6317bf372..4d4be3838 100644
--- a/numba/cuda/cudadecl.py
+++ b/numba/cuda/cudadecl.py
@@ -368,50 +368,6 @@ def _genfp16_binary(l_key):
 
     return Cuda_fp16_binary
 
-# If multiple ConcreteTemplates provide typing for a single function, then
-# function resolution will pick the first compatible typing it finds even if it
-# involves inserting a cast that would be considered undesirable (in this
-# specific case, float16s could be cast to float32s for comparisons).
-#
-# To work around this, we instead use an AbstractTemplate that implements
-# exactly the casting logic that we desire. The AbstractTemplate gets
-# considered in preference to ConcreteTemplates during typing.
-#
-# This is tracked as Issue #7863 (https://github.com/numba/numba/issues/7863) -
-# once this is resolved it should be possible to replace this AbstractTemplate
-# with a ConcreteTemplate to simplify the logic.
-
-
-def _genfp16_binary_operator(l_key):
-    @register_global(l_key)
-    class Cuda_fp16_binary2(AbstractTemplate):
-        key = l_key
-
-        def generic(self, args, kws):
-            assert not kws
-
-            if len(args) == 2 and \
-                    (args[0] == types.float16 or args[1] == types.float16):
-                if (args[0] == types.float16):
-                    convertible = self.context.can_convert(args[1], args[0])
-                else:
-                    convertible = self.context.can_convert(args[0], args[1])
-
-                # We allow three cases here on binary operators:
-                #
-                # 1. Binary operator on two fp16 args - Conversion.exact
-                # 2. Binary op where a fp16 can be promoted
-                #    - Conversion.promote
-                # 3. Binary op of fp16 and int8, where int8 is converted to
-                #    fp16 - Conversion.safe
-                if (convertible == Conversion.exact) or \
-                   (convertible == Conversion.promote) or \
-                   (convertible == Conversion.safe):
-                    return signature(types.float16, types.float16,
-                                     types.float16)
-
-    return Cuda_fp16_binary2
-
 
 @register_global(float)
 class Float(AbstractTemplate):
@@ -448,9 +404,9 @@ def _genfp16_binary_comparison(l_key):
 # This is tracked as Issue #7863 (https://github.com/numba/numba/issues/7863) -
 # once this is resolved it should be possible to replace this AbstractTemplate
 # with a ConcreteTemplate to simplify the logic.
-def _genfp16_comparison_operator(l_key):
+def _fp16_binary_operator(l_key, retty):
     @register_global(l_key)
-    class Cuda_fp16_operator_cmp(AbstractTemplate):
+    class Cuda_fp16_operator(AbstractTemplate):
         key = l_key
 
         def generic(self, args, kws):
@@ -474,7 +430,17 @@ def _genfp16_comparison_operator(l_key):
                 if (convertible == Conversion.exact) or \
                    (convertible == Conversion.promote) or \
                    (convertible == Conversion.safe):
-                    return signature(types.b1, types.float16, types.float16)
+                    return signature(retty, types.float16, types.float16)
+
+    return Cuda_fp16_operator
+
+
+def _genfp16_comparison_operator(op):
+    return _fp16_binary_operator(op, types.b1)
+
+
+def _genfp16_binary_operator(op):
+    return _fp16_binary_operator(op, types.float16)
 
 
 Cuda_hadd = _genfp16_binary(cuda.fp16.hadd)
```

Here instead of duplicating everything we just specialize the returned class depending on the return type instead.

Do you see a reason not to do it this way?

testhound replied:
@gmarkall I like this change, very elegant.

gmarkall reviewed (changes requested):
Thanks for the updates - there are some further comments on the diff to look at. I'd not previously looked at the tests (as I'd been holding on in case @isVoid wanted to look at them based on his comments about testing inplace ops, but I think he's not been active on Numba recently), so I've now looked over them.

testhound commented:
gpuci run tests

testhound commented:
@gmarkall code updated with all your comments.

gmarkall reviewed (changes requested):
Thanks for the changes. I have a follow-up question about the copy/pasted code.

testhound commented:
gpuci run tests

testhound commented:
@gmarkall all requested changes made.

gmarkall reviewed (dismissed):
Many thanks for the changes. I think this now looks good.

@esc / @stuartarchibald Could this have a CUDA buildfarm run please?

testhound commented:
gpuci run tests

esc commented:
`Build numba_smoketest_cuda_yaml_154 has started`

gmarkall commented:
@testhound I understand that `numba.cuda.tests.cudapy.test_operator.TestOperatorModule.test_fp16_comparison` failed on the buildfarm because this PR accidentally removes the decorator to skip it when the CC is less than 5.3 Could you add it back please?

testhound commented:
gpuci run tests

testhound commented:
@gmarkall I added the skip_unless_cc_53 back to the testcase.

gmarkall commented:
@esc Could this have another buildfarm run please?

esc commented:
> @esc Could this have another buildfarm run please?

`numba_smoketest_cuda_yaml_156`

esc commented:
> > @esc Could this have another buildfarm run please?
> 
> `numba_smoketest_cuda_yaml_156`

So, I tested `` on the Anaconda build farm and still received the following errors on Windows.

@testhound could you take a look, please?

```
======================================================================
ERROR: test_float_to_float (numba.cuda.tests.cudapy.test_casting.TestCasting) (fromty=<class 'numpy.float16'>, toty=<class 'numpy.float16'>)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\cudadrv\driver.py", line 2705, in add_ptx
    driver.cuLinkAddData(self.handle, enums.CU_JIT_INPUT_PTX,
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\cudadrv\driver.py", line 320, in safe_cuda_api_call
    self._check_ctypes_error(fname, retcode)
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\cudadrv\driver.py", line 388, in _check_ctypes_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [218] Call to cuLinkAddData results in UNKNOWN_CUDA_ERROR

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\tests\cudapy\test_casting.py", line 210, in test_float_to_float
    np.testing.assert_allclose(cfunc(12.3),
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\tests\cudapy\test_casting.py", line 111, in wrapper_fn
    cuda_wrapper_fn[1, 1](argarray, resarray)
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 505, in __call__
    return self.dispatcher.call(args, self.griddim, self.blockdim,
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 639, in call
    kernel = _dispatcher.Dispatcher._cuda_call(self, *args)
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 647, in _compile_for_args
    return self.compile(tuple(argtypes))
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 849, in compile
    kernel.bind()
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 178, in bind
    self._codelibrary.get_cufunc()
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\codegen.py", line 208, in get_cufunc
    cubin = self.get_cubin(cc=device.compute_capability)
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\codegen.py", line 181, in get_cubin
    linker.add_ptx(ptx.encode())
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\cudadrv\driver.py", line 2708, in add_ptx
    raise LinkerError("%s\n%s" % (e, self.error_log))
numba.cuda.cudadrv.driver.LinkerError: [218] Call to cuLinkAddData results in UNKNOWN_CUDA_ERROR
ptxas application ptx input, line 56; error   : Feature 'f16 arithemetic and compare instructions' requires .target sm_53 or higher
ptxas fatal   : Ptx assembly aborted due to errors

======================================================================
ERROR: test_float_to_float (numba.cuda.tests.cudapy.test_casting.TestCasting) (fromty=<class 'numpy.float16'>, toty=<class 'numpy.float32'>)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\cudadrv\driver.py", line 2705, in add_ptx
    driver.cuLinkAddData(self.handle, enums.CU_JIT_INPUT_PTX,
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\cudadrv\driver.py", line 320, in safe_cuda_api_call
    self._check_ctypes_error(fname, retcode)
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\cudadrv\driver.py", line 388, in _check_ctypes_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [218] Call to cuLinkAddData results in UNKNOWN_CUDA_ERROR

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\tests\cudapy\test_casting.py", line 210, in test_float_to_float
    np.testing.assert_allclose(cfunc(12.3),
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\tests\cudapy\test_casting.py", line 111, in wrapper_fn
    cuda_wrapper_fn[1, 1](argarray, resarray)
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 505, in __call__
    return self.dispatcher.call(args, self.griddim, self.blockdim,
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 639, in call
    kernel = _dispatcher.Dispatcher._cuda_call(self, *args)
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 647, in _compile_for_args
    return self.compile(tuple(argtypes))
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 849, in compile
    kernel.bind()
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 178, in bind
    self._codelibrary.get_cufunc()
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\codegen.py", line 208, in get_cufunc
    cubin = self.get_cubin(cc=device.compute_capability)
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\codegen.py", line 181, in get_cubin
    linker.add_ptx(ptx.encode())
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\cudadrv\driver.py", line 2708, in add_ptx
    raise LinkerError("%s\n%s" % (e, self.error_log))
numba.cuda.cudadrv.driver.LinkerError: [218] Call to cuLinkAddData results in UNKNOWN_CUDA_ERROR
ptxas application ptx input, line 57; error   : Feature 'f16 arithemetic and compare instructions' requires .target sm_53 or higher
ptxas fatal   : Ptx assembly aborted due to errors

======================================================================
ERROR: test_float_to_float (numba.cuda.tests.cudapy.test_casting.TestCasting) (fromty=<class 'numpy.float16'>, toty=<class 'numpy.float64'>)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\cudadrv\driver.py", line 2705, in add_ptx
    driver.cuLinkAddData(self.handle, enums.CU_JIT_INPUT_PTX,
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\cudadrv\driver.py", line 320, in safe_cuda_api_call
    self._check_ctypes_error(fname, retcode)
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\cudadrv\driver.py", line 388, in _check_ctypes_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [218] Call to cuLinkAddData results in UNKNOWN_CUDA_ERROR

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\tests\cudapy\test_casting.py", line 210, in test_float_to_float
    np.testing.assert_allclose(cfunc(12.3),
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\tests\cudapy\test_casting.py", line 111, in wrapper_fn
    cuda_wrapper_fn[1, 1](argarray, resarray)
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 505, in __call__
    return self.dispatcher.call(args, self.griddim, self.blockdim,
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 639, in call
    kernel = _dispatcher.Dispatcher._cuda_call(self, *args)
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 647, in _compile_for_args
    return self.compile(tuple(argtypes))
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 849, in compile
    kernel.bind()
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 178, in bind
    self._codelibrary.get_cufunc()
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\codegen.py", line 208, in get_cufunc
    cubin = self.get_cubin(cc=device.compute_capability)
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\codegen.py", line 181, in get_cubin
    linker.add_ptx(ptx.encode())
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\cudadrv\driver.py", line 2708, in add_ptx
    raise LinkerError("%s\n%s" % (e, self.error_log))
numba.cuda.cudadrv.driver.LinkerError: [218] Call to cuLinkAddData results in UNKNOWN_CUDA_ERROR
ptxas application ptx input, line 56; error   : Feature 'f16 arithemetic and compare instructions' requires .target sm_53 or higher
ptxas fatal   : Ptx assembly aborted due to errors

======================================================================
ERROR: test_int_to_float (numba.cuda.tests.cudapy.test_casting.TestCasting) (toty=<class 'numpy.float16'>)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\cudadrv\driver.py", line 2705, in add_ptx
    driver.cuLinkAddData(self.handle, enums.CU_JIT_INPUT_PTX,
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\cudadrv\driver.py", line 320, in safe_cuda_api_call
    self._check_ctypes_error(fname, retcode)
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\cudadrv\driver.py", line 388, in _check_ctypes_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [218] Call to cuLinkAddData results in UNKNOWN_CUDA_ERROR

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\tests\cudapy\test_casting.py", line 167, in test_int_to_float
    self.assertEqual(cfunc(321), pyfunc(321))
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\tests\cudapy\test_casting.py", line 111, in wrapper_fn
    cuda_wrapper_fn[1, 1](argarray, resarray)
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 505, in __call__
    return self.dispatcher.call(args, self.griddim, self.blockdim,
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 639, in call
    kernel = _dispatcher.Dispatcher._cuda_call(self, *args)
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 647, in _compile_for_args
    return self.compile(tuple(argtypes))
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 849, in compile
    kernel.bind()
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\dispatcher.py", line 178, in bind
    self._codelibrary.get_cufunc()
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\codegen.py", line 208, in get_cufunc
    cubin = self.get_cubin(cc=device.compute_capability)
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\codegen.py", line 181, in get_cubin
    linker.add_ptx(ptx.encode())
  File "F:\ci_envs\64\Miniconda3\envs\testenv_9ce6cbd8-4fe8-4839-bd6c-bfb4871ad7e2\lib\site-packages\numba\cuda\cudadrv\driver.py", line 2708, in add_ptx
    raise LinkerError("%s\n%s" % (e, self.error_log))
numba.cuda.cudadrv.driver.LinkerError: [218] Call to cuLinkAddData results in UNKNOWN_CUDA_ERROR
ptxas application ptx input, line 59; error   : Feature 'f16 arithemetic and compare instructions' requires .target sm_53 or higher
ptxas fatal   : Ptx assembly aborted due to errors

----------------------------------------------------------------------
Ran 1360 tests in 468.759s

FAILED (errors=4, skipped=82, expected failures=10)
Parallel: 351. Serial: 1009
Command exited with return code 1
```

testhound commented:
@esc thank you for the feedback, looking into the failures.

testhound commented:
@esc can I request another buildfarm run? I fixed the failing tests by skipping if the compute capability requirements are not met. I am confused as to why these tests just started failing, since this code has been in main for quite some time.

testhound commented:
@gmarkall FYI: I updated the PR to fix tests that should be skipped.

gmarkall commented:
>  I am confused as to why these tests just started failing, since this code has been in main for quite some time.

It suggests that there is something wrong with these tests that we never picked up on before.

stuartarchibald commented on numba/cuda/tests/nocuda/test_function_resolution.py:
```diff
@@ -0,0 +1,37 @@
+            typingctx.refresh()
+            fnty = typingctx.resolve_value_type(op)
+            out = typingctx.resolve_function_type(fnty, (fp16, fp16), {})
+            if out != typing.signature(fp16, fp16, fp16):
+                raise AssertionError(out)
```
Should this be `self.assertEqual(out, typing.signature(fp16, fp16, fp16), msg=str(out))` or similar? (same for the other test in this class).

gmarkall replied:
Yes.

testhound commented:
gpuci run tests

testhound commented:
@stuartarchibald and @gmarkall code update per your request to use `assertEqual`.

testhound commented:
gpuci run tests

gmarkall reviewed (approved):
Looks good to me. Could this have another buildfarm run please @stuartarchibald ?

stuartarchibald reviewed (approved):
Thanks for the patch and fixes.

esc commented:
I ran this on the farm as `numba_smoketest_cuda_yaml_160` and it was fine for most tasks. Our Linux builder can't run with `cudatoolkit` 11.3 so I couldn't test that -- but the Windows one can and that passed.
